Sources/NIOHPACK/HuffmanTables.swift

```diff
@@ -0,0 +1,5404 @@
+ included in all copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+typealias HuffmanDecodeEntry = (state: UInt8, flags: HuffmanDecoderFlags, sym: UInt8)
+
+internal struct HuffmanDecoderFlags : OptionSet
+{
+    var rawValue: UInt8
+    
+    static let none     = HuffmanDecoderFlags(rawValue: 0b000)
+    static let accepted = HuffmanDecoderFlags(rawValue: 0b001)
+    static let symbol   = HuffmanDecoderFlags(rawValue: 0b010)
+    static let failure  = HuffmanDecoderFlags(rawValue: 0b100)
+}
+
+/**
+ This was described nicely by `@Lukasa` in his Python implementation:
+ 
+ The essence of this approach is that it builds a finite state machine out of
+ 4-bit nybbles of Huffman coded data. The input function passes 4 bits worth of
+ data to the state machine each time, which uses those 4 bits of data along with
+ the current accumulated state data to process the data given.
+
+ For the sake of efficiency, the in-memory representation of the states,
+ transitions, and result values of the state machine are represented as a long
+ list containing three-tuples. This list is enormously long, and viewing it as
+ an in-memory representation is not very clear, but it is laid out here in a way
+ that is intended to be *somewhat* more clear.
+
+ Essentially, the list is structured as 256 collections of 16 entries (one for
+ each nybble) of three-tuples. Each collection is called a "node", and the
+ zeroth collection is called the "root node". The state machine tracks one
+ value: the "state" byte.
+
+ For each nybble passed to the state machine, it first multiplies the "state"
+ byte by 16 and adds the numerical value of the nybble. This number is the index
+ into the large flat list.
+
+ The three-tuple that is found by looking up that index consists of three
+ values:
+
+ - a new state value, used for subsequent decoding
+ - a collection of flags, used to determine whether data is emitted or whether
+ the state machine is complete.
+ - the byte value to emit, assuming that emitting a byte is required.
+
+ The flags are consulted, if necessary a byte is emitted, and then the next
+ nybble is used. This continues until the state machine believes it has
+ completely Huffman-decoded the data.
+
+ This approach has relatively little indirection, and therefore performs
+ relatively well, particularly on implementations like PyPy where the cost of
+ loops at the Python-level is not too expensive. The total number of loop
+ iterations is 4x the number of bytes passed to the decoder.
```

Comment:
Hah, we can probably drop this last paragraph. ;)

Reply:
I thought it was fairly illustrative, actually 😊